toolboxes/core/gpu/cuNDArray_blas.cu

```diff
@@ -1,312 +1,509 @@
+}
 
-  template<> cublasStatus_t cublas_amin<double_complext>(cublasHandle_t hndl, int n,const double_complext *x, int incx, int *result){
-    return cublasIzamin(hndl,n, (const cuDoubleComplex* ) x,incx,result);
-  }
+template <>
+cublasStatus_t cublas_amin<double_complext>(cublasHandle_t hndl, int n, const double_complext* x, int incx,
+                                            int* result) {
+    return cublasIzamin(hndl, n, (const cuDoubleComplex*)x, incx, result);
+}
 
-  //AMAX
-  //
+// AMAX
+//
 
-  template<> cublasStatus_t cublas_amax<float>(cublasHandle_t hndl, int n,const float *x, int incx, int *result){
-    return cublasIsamax(hndl,n,x,incx,result);
-  }
+template <> cublasStatus_t cublas_amax<float>(cublasHandle_t hndl, int n, const float* x, int incx, int* result) {
+    return cublasIsamax(hndl, n, x, incx, result);
+}
 
-  template<> cublasStatus_t cublas_amax<double>(cublasHandle_t hndl, int n,const double *x, int incx, int *result){
-    return cublasIdamax(hndl,n,x,incx,result);
-  }
+template <> cublasStatus_t cublas_amax<double>(cublasHandle_t hndl, int n, const double* x, int incx, int* result) {
+    return cublasIdamax(hndl, n, x, incx, result);
+}
 
-  template<> cublasStatus_t cublas_amax<float_complext>(cublasHandle_t hndl, int n,const float_complext *x, int incx, int *result){
-    return cublasIcamax(hndl,n, (const cuComplex* ) x,incx,result);
-  }
+template <>
+cublasStatus_t cublas_amax<float_complext>(cublasHandle_t hndl, int n, const float_complext* x, int incx, int* result) {
+    return cublasIcamax(hndl, n, (const cuComplex*)x, incx, result);
+}
 
-  template<> cublasStatus_t cublas_amax<double_complext>(cublasHandle_t hndl, int n,const double_complext *x, int incx, int *result){
-    return cublasIzamax(hndl,n, (const cuDoubleComplex* ) x,incx,result);
-  }
+template <>
+cublasStatus_t cublas_amax<double_complext>(cublasHandle_t hndl, int n, const double_complext* x, int incx,
+                                            int* result) {
+    return cublasIzamax(hndl, n, (const cuDoubleComplex*)x, incx, result);
+}
 
-  template<class T> typename realType<T>::Type nrm2( cuNDArray<T> *arr )
-  {
-    if( arr == 0x0 )
-      throw std::runtime_error("Gadgetron::nrm2(): Invalid input array");
+template <class T> typename realType<T>::Type nrm2(cuNDArray<T>* arr) {
+    if (arr == 0x0)
+        throw std::runtime_error("Gadgetron::nrm2(): Invalid input array");
 
     int device = cudaDeviceManager::Instance()->getCurrentDevice();
     typedef typename realType<T>::Type REAL;
     REAL ret;
 
-    CUBLAS_CALL(cublas_nrm2<T>( cudaDeviceManager::Instance()->lockHandle(device), 
-                                (int)arr->get_number_of_elements(), arr->get_data_ptr(), 1, &ret));
+    // If number of elements in the array is greater than INT_MAX break it up and perform calculations this is done to
+    // support large data arrays
+
+    int num_splits = arr->get_number_of_elements() / INT_MAX + 1;
```

Comment:
Feels like there is a pattern of a "batched" cublas call that could be broken out here so we handling this splitting logic in one place and just calling into this batched cublas wrapper with a function pointer to the operation we want to do?